### PR TITLE
SpNavbar and Keplr connection

### PR DESCRIPTION
**Pull request by Matthews3301**
Navbar component which allows a user to connect to Keplr account.

https://github.com/tendermint/vue/issues/136

* Detect if Keplr is available
* Add new chain to Keplr
* Set is as active
* Handle connection cancellations and errors
* Retrieve account name and address
* Save to vuex store

### CODE REVIEW

**Review by marinhoarthur — CHANGES_REQUESTED** [review]
Points:

- Lack of components that are being used (`SpButton`)
- Not able to connect to Keplr (error)
- `SpNavbar` wrapping too many responsibilities. Split Keplr stuff into a dedicated component (`SpAcc` or smth)

And minors:

- Remove `depsLoaded` and overall cleaning
- Let's get closer to each other as far as code style goes #158 (refer to `SpTx`)
- Since icons were put into separate files let's use our agreed new folder structure (e.g.: src/components/ui)

**Comment by Matthews3301** [comment]
@marinhoarthur what error are you getting when connecting to Keplr? What does it say in the console?

**Comment by Matthews3301** [comment]
@marinhoarthur please check commit e89c2f4 for fixes related to your points.

**Comment by Matthews3301** [comment]
@marinhoarthur not sure why this is happening but all indenting is using tabs... Would it not be better to use double spaces? It looks better in GitHub preview.

**Comment by marinhoarthur** [comment]
Pushed: 

- Took account stuff out of `SpNavbar` and put in `SpAcc` instead
- Moved the icons files around because rollup had a problem while loading those (we can tackle this later!)
- The `useKeplr` composable was updating component specific reactive stuff. Don't think this is desired. Why don't we use callbacks passed to the `connectToKeplr` (e.g.: onConnect, onError)?

File: packages/vue/src/components/SpAcc/SpAcc.vue [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+		SpAccountDropdown,
+		SpKeplrIcon,
+		SpWarningIcon,
+		SpExternalArrowIcon,
+		SpChevronDownIcon
+	},
+
+	setup() {
+		// state
+		const state = reactive({
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpAcc/SpAcc.vue]: Let's use [this approach](https://github.com/tendermint/vue/blob/07ad9b921f73568febf1d3b29bad0d833c08add1/packages/vue/src/components/SpTx/SpTx.vue#L309) to increase typing capabilities?

File: packages/vue/src/components/SpAcc/SpAcc.vue [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+		// $s
+		const $s = useStore()
+
+		// computed
+
+		// methods
+		const updateWalletData = async (chainId: string) => {
+			const walletParams = await window.keplr.getKey(chainId)
+			const offlineSigner = window.keplr.getOfflineSigner(chainId)
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpAcc/SpAcc.vue]: Since we now have `useKeplr` composable, better to move all keplr related stuff in there, right?

**Comment by marinhoarthur** [comment]
> @marinhoarthur not sure why this is happening but all indenting is using tabs... Would it not be better to use double spaces? It looks better in GitHub preview.

probably some code formatting tuning...

File: packages/vue/src/components/SpAcc/SpAcc.vue [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    const updateWalletData = async (chainId: string) => {
+      const walletParams = await window.keplr.getKey(chainId)
+      const offlineSigner = window.keplr.getOfflineSigner(chainId)
+      await $s.dispatch('common/wallet/connectWithKeplr', offlineSigner)
+      state.connectedWallet = {
+        name: walletParams.name,
+        address: walletParams.bech32Address
+      }
+    }
+    const autoConnectToChain = async (chainId: string) => {
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpAcc/SpAcc.vue]: what's the purpose of auto connecting a chain?

File: packages/vue/src/components/SpAcc/SpAcc.vue [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+			state.modalPage = 'error'
+    }
+    const connectToKeplr = () => {
+			state.modalPage = 'connecting'
+      connectToKeplrAPI(onKeplrConnect, onKeplrError)
+    }
+
+    // lh
+    onBeforeMount(async () => {
+      const chainId = $s.getters['common/env/chainId']
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpAcc/SpAcc.vue]: this should be a computed prop

File: packages/vue/src/components/SpAcc/SpAcc.vue [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+      if (chainId) {
+        try {
+          await updateWalletData(chainId)
+        } catch (e) {
+          console.log('Keplr not connected')
+        }
+      }
+    }
+    const onKeplrConnect = () => {
+      updateWalletData($s.getters['common/env/chainId'])
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpAcc/SpAcc.vue]: we need to set a listener to solve this #151

File: packages/vue/src/components/SpAcc/SpAcc.vue [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+    // computed
+
+    // composables
+    let { connectToKeplrAPI, keplrAvailable } = useKeplr($s)
+
+    // methods
+    const updateWalletData = async (chainId: string) => {
+      const walletParams = await window.keplr.getKey(chainId)
+      const offlineSigner = window.keplr.getOfflineSigner(chainId)
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpAcc/SpAcc.vue]: like said before, keplr api should be exposed by `useKeplr`

File: packages/vue/src/components/SpAcc/SpAcc.vue [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+      await $s.dispatch('common/wallet/connectWithKeplr', offlineSigner)
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpAcc/SpAcc.vue]: ideally let's have an action block listing all store actions

File: packages/vue/src/components/SpAccountDropdown/SpAccountDropdown.vue [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+	},
+	props: {
+		connectedWallet: {
+			type: Object,
+			required: true
+		}
+	},
+	setup(_, { emit }) {
+		const store = useStore()
+
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpAccountDropdown/SpAccountDropdown.vue]: let's keep this component simple and just emit `disconnect`, leaving spacc with the `signOut` dispatching job

File: packages/vue/src/composables/useKeplr.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
+import { computed } from 'vue'
+import { Store } from 'vuex'
+
+export interface Amount {
```
Comment (marinhoarthur) [inline packages/vue/src/composables/useKeplr.ts]: these two interfaces are already implemented & exposed by `utils/interfaces`

File: packages/vue/src/composables/useKeplr.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+              coinDecimals: 0
+            }
+            y.coinDenom = x.denom.toUpperCase()
+            y.coinMinimalDenom = x.denom
+            y.coinDecimals = 0
+            return y
+          }),
+          coinType: 118,
+          gasPriceStep: {
+            low: 0.01,
```
Comment (marinhoarthur) [inline packages/vue/src/composables/useKeplr.ts]: for the future: make these dynamic

File: packages/vue/src/composables/useKeplr.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            const y: AmountWithMeta = {
+              amount: '0',
+              denom: '',
+              coinDenom: '',
+              coinMinimalDenom: '',
+              coinDecimals: 0
+            }
+            y.coinDenom = x.denom.toUpperCase()
+            y.coinMinimalDenom = x.denom
+            y.coinDecimals = 0
```
Comment (marinhoarthur) [inline packages/vue/src/composables/useKeplr.ts]: for the future: make these dynamic

File: packages/vue/src/components/SpNavbar/SpNavbar.vue [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    },
+    selectedLink: {
+      type: String,
+      default: 'Portfolio'
+    }
+  },
+
+  setup(props) {
+    const data = reactive({
+      links: props.links,
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpNavbar/SpNavbar.vue]: reactive over reactive, why?

**Comment by marinhoarthur** [comment]
Just [pushed](https://github.com/tendermint/vue/pull/156/commits/00fee5c3d95b2af248b74a3672306c2784ab71fc) changes tackling most of my comments.

@Matthews3301 could you please go over the comments and see what's missing, specially `SpNavbar` _reactive over reactive_ and lack of typing on `SpAcc`'s state.

File: packages/vue/src/components/SpAcc/SpAcc.vue [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+		SpAccountDropdown,
+		SpKeplrIcon,
+		SpWarningIcon,
+		SpExternalArrowIcon,
+		SpChevronDownIcon
+	},
+
+	setup() {
+		// state
+		let state = reactive({
```
Comment (marinhoarthur) [inline packages/vue/src/components/SpAcc/SpAcc.vue]: Let's type state as in [SpTx](https://github.com/tendermint/vue/blob/07ad9b921f73568febf1d3b29bad0d833c08add1/packages/vue/src/components/SpTx/SpTx.vue#L309) please!

**Review by marinhoarthur — CHANGES_REQUESTED** [review]
See comments

**Comment by marinhoarthur** [comment]
solves #151

File: packages/vue/src/composables/useKeplr.ts (hunk 10 above)
Comment (marinhoarthur) [reply packages/vue/src/composables/useKeplr.ts]: related to #153

**Comment by marinhoarthur** [comment]
`SpAcc` is a good candidate to have an enum as ui state as in [here](https://github.com/tendermint/vue/blob/07ad9b921f73568febf1d3b29bad0d833c08add1/packages/vue/src/components/SpTx/SpTx.vue#L286). Having multiple booleans getting together to express current ui state is hard to maintain.

**Review by marinhoarthur — APPROVED** [review]

**Pull request merged**